File: src/main/java/com/portofino/realtrainmodunofficial/network/ConfigureMarkerPayload.java
package com.portofino.realtrainmodunofficial.network;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficial;
import com.portofino.realtrainmodunofficial.blockentity.MarkerBlockEntity;
import io.netty.buffer.ByteBuf;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.neoforge.network.handling.IPayloadContext;

/** レンチ設定GUI → サーバ: マーカーのアンカー(曲線)・カント(傾き)を更新する。 */
public record ConfigureMarkerPayload(BlockPos pos, float anchorYaw, float anchorPitch,
                                     float anchorLengthHorizontal, float anchorLengthVertical,
                                     float cantCenter, float cantEdge, float cantRandom) implements CustomPacketPayload {
    public static final Type<ConfigureMarkerPayload> TYPE = new Type<>(
        ResourceLocation.fromNamespaceAndPath(RealTrainModUnofficial.MODID, "configure_marker")
    );

    // composite は最大6要素のため手動コーデックにする (フィールドが8個)。
    public static final StreamCodec<ByteBuf, ConfigureMarkerPayload> STREAM_CODEC = StreamCodec.of(
        (buf, p) -> {
            BlockPos.STREAM_CODEC.encode(buf, p.pos);
            buf.writeFloat(p.anchorYaw);
            buf.writeFloat(p.anchorPitch);
            buf.writeFloat(p.anchorLengthHorizontal);
            buf.writeFloat(p.anchorLengthVertical);
            buf.writeFloat(p.cantCenter);
            buf.writeFloat(p.cantEdge);
            buf.writeFloat(p.cantRandom);
        },
        buf -> new ConfigureMarkerPayload(
            BlockPos.STREAM_CODEC.decode(buf),
            buf.readFloat(), buf.readFloat(), buf.readFloat(), buf.readFloat(),
            buf.readFloat(), buf.readFloat(), buf.readFloat()
        )
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnServer(ConfigureMarkerPayload payload, IPayloadContext context) {
        context.enqueueWork(() -> {
            if (!(context.player() instanceof ServerPlayer player)) {
                return;
            }
            if (!(player.level().getBlockEntity(payload.pos()) instanceof MarkerBlockEntity marker)) {
                return;
            }
            marker.configure(payload.anchorYaw(), payload.anchorPitch(),
                payload.anchorLengthHorizontal(), payload.anchorLengthVertical(),
                payload.cantCenter(), payload.cantEdge(), payload.cantRandom());
            player.displayClientMessage(Component.literal("マーカー設定を更新しました"), true);
        });
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/ConfigureMarkerPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.blockentity.MarkerBlockEntity
import io.netty.buffer.ByteBuf
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.handling.IPayloadContext

/** Wrench config GUI to server: updates marker anchor curve and cant values. */
@JvmRecord
data class ConfigureMarkerPayload(
    val pos: BlockPos,
    val anchorYaw: Float,
    val anchorPitch: Float,
    val anchorLengthHorizontal: Float,
    val anchorLengthVertical: Float,
    val cantCenter: Float,
    val cantEdge: Float,
    val cantRandom: Float,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<ConfigureMarkerPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "configure_marker")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, ConfigureMarkerPayload> = StreamCodec.of(
            { buf, payload ->
                BlockPos.STREAM_CODEC.encode(buf, payload.pos)
                buf.writeFloat(payload.anchorYaw)
                buf.writeFloat(payload.anchorPitch)
                buf.writeFloat(payload.anchorLengthHorizontal)
                buf.writeFloat(payload.anchorLengthVertical)
                buf.writeFloat(payload.cantCenter)
                buf.writeFloat(payload.cantEdge)
                buf.writeFloat(payload.cantRandom)
            },
            { buf ->
                ConfigureMarkerPayload(
                    BlockPos.STREAM_CODEC.decode(buf),
                    buf.readFloat(),
                    buf.readFloat(),
                    buf.readFloat(),
                    buf.readFloat(),
                    buf.readFloat(),
                    buf.readFloat(),
                    buf.readFloat(),
                )
            },
        )

        @JvmStatic
        fun handleOnServer(payload: ConfigureMarkerPayload, context: IPayloadContext) {
            context.enqueueWork {
                val player = context.player() as? ServerPlayer ?: return@enqueueWork
                val marker = player.level().getBlockEntity(payload.pos) as? MarkerBlockEntity ?: return@enqueueWork
                marker.configure(
                    payload.anchorYaw,
                    payload.anchorPitch,
                    payload.anchorLengthHorizontal,
                    payload.anchorLengthVertical,
                    payload.cantCenter,
                    payload.cantEdge,
                    payload.cantRandom,
                )
                player.sendOverlayMessage(Component.literal("マーカー設定を更新しました"))
            }
        }
    }
}
